# CpGSelectionPipeline

This repository comprises scripts to select CpGs from bulk DNA methylation data for subsequent analysis with scTAM-seq. 

In general, three classes of CpGs are selected:
- Differentially Methylated CpGs (DMCs, in [RnBeads](RnBeads))
- Intermediately Methylated CpGs (IMCs, in [IMC](IMC))
- CpGs harboring Within-Sample Heterogeneity (WSH, in [WSH](WSH))

Further scripts, such as those for annotating selected sites, selecting constitutively methylated and unmethylated sites, and amplicons without an HhaI cutsite can be found in [scripts](scripts). Lastly, utilities to generate visualization of the results can be found in [plotting](plotting).

General configuration of the pipeline, such as the number of CpGs to be covered in the panel, can be provided in the file *config.yaml*.

# Collecting the data

Most parts of the selection pipeline use the [Cabezas-Wallscheid, 2014](https://doi.org/10.1016/j.stem.2014.07.005) dataset. We recommend to use a standard bisulfite sequencing data processing pipeline, such as [bismark](https://doi.org/10.1093/bioinformatics/btr167) for low-level data processing. Bismark files can be used directly as input to [RnBeads](https://doi.org/10.1186/s13059-019-1664-9), the generated bam-files can be used to compute the [PDR](https://doi.org/10.1016/j.ccell.2014.10.012) and [qFDRP](https://doi.org/10.1093/nar/gkaa120).

# Installing dependencies

The pipeline mainly depends on [RnBeads](https://rnbeads.org), which can be installed using:

```{r}
source('https://rnbeads.org/data/install.R')
```

Furthermore, the following R packages are required:
-yaml
-data.table
-argparse

To faciliate installing, we provide a yaml file for easy installation with conda/bioconda
 
```
conda create -f selection_pipeline.yml -n selection_pipeline
```

# Running the pipeline

The pipeline has been targeted to an HPC system operated by the SGE jobs scheduling system. For other HPC environments or a local installation, the scripts have to be changed accordingly.

For each of the categories of CpGs mentioned above, a dedicated script has been created:
- In [RnBeads](RnBeads), you'll have to point in the rnbOptions.xml and script.R file to the downloaded data directory and adjust the file according to your system. See the [RnBeads documentation](http://bioconductor.org/packages/release/bioc/vignettes/RnBeads/inst/doc/RnBeads.pdf) for further information in case you want to modify the analysis. The example sample_annotation.csv may have to be adapted according to the filenames of the downloaded files. Lastly, the script *select_reliable_DMRs.R* will select sites that are specifically methylated in each of the populations.
- To compute the WSH scores PDR and qFDRP, the bam-files generated by bismark are leveraged. Script to compute the scores are available from [WSH](WSH).
- IMCs are defined as sites with intermediate methylation in the HSCs and low PDR scores. They can be determined using the *find_IMCs_low_PDR.R* in the [IMC](IMC) folder.
- Variably methylated sites are selected based on those that are not IMCs (i.e. have high PDR), but also high qFDRP scores. For this, the script *filter_qFDRP_results.R* is responsible.
- To include control regions (non-digested, constitutively methylated, constitutively unmethylated), corresponding scripts have been generated in [scripts](scripts).

After successful completion of all of these scripts, the final panel can be generated and uploaded to the Mission Bio Tapestri Designer tool. For this, the script *generate_final_panel.R* in [scripts](scripts) is targeted for. 

# Contact

You can contact [Michael Scherer](mailto:michael.scherer@crg.eu) for questions and suggestions.
